foundry/https_upload.py

```diff
@@ -0,0 +1,238 @@
+"""
+https_upload.py
+
+Private utility methods to upload files and/or folders to Globus using HTTPS instead of Globus Transfer.
+
+Authors:
+    Aristana Scourtas
+    Isaac Darling
+
+Last modified 12/21/22 by Isaac Darling
+"""
+
+import logging
+import os
+import urllib
+from requests import put, Response
+from typing import Any, Tuple, Dict, List
+from uuid import uuid4
+
+from globus_sdk import AuthClient, TransferAPIError
+
+
+logger = logging.getLogger(__name__)
+
+
+def upload_to_endpoint(auths: dict, local_data_path: str, endpoint_id: str = "82f1b5c6-6e9b-11e5-ba47-22000b92c6ec",
+                       dest_parent: str = None, dest_child: str = None) -> Tuple[str, str]:
+    """Upload local data to a Globus endpoint using HTTPS PUT requests. Data can be a folder or an individual file.
+        Note that the ACL rule created in this method must later be deleted after the dataset is submitted to MDF.
+    Args:
+        auths (dict): the required authorizers
```

Comment:
Would it be possible to pass the TransferClient and AuthClient as inputs? That way, users don't need to know how the dictionary must be structures (e.g.,, 'transfer' -> TransferClient)

Reply:
yes I think that makes sense! good catch, I didn't realize how much unnecessary unpacking happens downstream

Reply:
Although actually @WardLT , I don't know if it would make that much of a difference -- currently we need the AuthClient,  the TransferClient, and the authorizer for the NCSA scope (see this line). So to pass all 3 of those, I think I'd want to bundle them in an object anyways, which means we'll still need to reference by key. 

What do you think? 

If we do keep the object passing, I could write more in the docstrings so that the user knows what keys map to what. And perhaps make it a smaller object than full `auths`

Reply:
I think this is more of an aesthetic decision then. I prefer having many arguments for a command so that the docstring is simpler, an IDE can help you catch when an argument is wrong, and you avoid having to write your own error messages for badly-formatted dictionaries. (e.g., `f(auth={'trnfser': client})` would not be detectable by an IDE, but `f(trnsfer=client)` would and would also throw an error in Python) 

That said, I'd back your play if you choose to keep the "put it all in a dictionary" approach

Reply:
Perhaps a `dataclass` or `NamedTuple` that holds the auth clients rather than a dictionary would be a middle ground: you could keep all of the authorizers together, but you would put up more guiderails of what the "authorizer collection" can contain.

Reply:
these are great points! I'm going to think about it a bit and do some googling. I really like everything you raised here. 

My goal was readability and ease of invoking a function. 

I like the idea of a NamedTuple

Reply:
Sounds good. Are you thinking still that the dictionary is the right object?

Reply:
I actually liked the NamedTuple best! felt like the best balance of user-friendly and lightweight (dataclass is thorough but felt like overkill, dict felt less user friendly since you need to know keys).